src/scheduler/pbs_sched.c

```diff
@@ -792,45 +506,400 @@ are_we_primary()
 }
 
 /**
- * @brief
- * 		the entry point of the pbs_sched.
+ * @brief close connections to given server and free its struct
+ *
+ * @param[in] - sconn - pointer to server struct to close
+ *
+ * @return void
+ */
+static void
+close_server(sched_svrconn *sconn)
+{
+	if (sconn) {
+		remove_ptr_from_array(servers, sconn);
+		if (count_array(servers) == 0) {
+			free(servers);
+			servers = NULL;
+		}
+		tpp_em_del_fd(poll_context, sconn->secondary_sock);
+		pbs_disconnect(sconn->primary_sock);
+		pbs_disconnect(sconn->secondary_sock);
+		log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO,
+			   msg_daemonname, "Disconnected from server %s", sconn->svrhost);
+		free(sconn->svrhost);
+		free(sconn);
+	}
+}
+
+/**
+ * @brief close connections to all servers
+ *
+ * @return void
+ */
+static void
+close_servers(void)
+{
+	while (count_array(servers))
+		close_server(servers[0]);
```

Comment:
Do we need to have a while loop to free servers here. Why cannot we make use of free_ptr_array() function.

Reply:
free_ptr_array() just frees array but doesn't free anything in ptr (aka entries of array).

Reply:
It also frees the individual elements of the array provided the members of the individual elements do not contain pointers. In our case we have to free svrhost of sch_svrconn structure in which case we cannot directly use it. But regardless we can follow the same style.

Better not to call count_array function every time in a loop rather can we simply do like the below which avoids calling count_array() function every iteration of the loop.
```
for (i = 0; servers[i] != NULL; i++)
	close_server(servers[i]);
free(servers);
```

Reply:
I had initially used same only, but it give me leak and other problems like invalid read while accessing server[i], as we have to remove give sconn ptr from servers inside close_server(), so here servers[i] will change at every call to close_server().

Also since no. servers won't be too much so count_array won't be much performance issue.

I can get rid of all this by simply using link list, but as @bhroam said sched doesn't use it.

Reply:
Okay. Just another thought. How about if we do the following. If we follow this the above style of code works fine. What do you think ?
```
static void
close_server(sched_svrconn *sconn, int remove_entry)
{
	if (sconn) {
		if (remove_entry)
			remove_ptr_from_array(servers, sconn);
		if (count_array(servers) == 0) {
			free(servers);
			servers = NULL;
		}
		tpp_em_del_fd(poll_context, sconn->secondary_sock);
		pbs_disconnect(sconn->primary_sock);
		pbs_disconnect(sconn->secondary_sock);
		log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO,
			   msg_daemonname, "Disconnected from server %s", sconn->svrhost);
		free(sconn->svrhost);
		free(sconn);
	}

}```

Wherever we want to close all servers i.e. close_servers() function call it as
	close_servers(servers[i], 0)
In all other places where only one server needs to be closed call it as
	close_servers(svr_entry, 1)

Reply:
@suresh-thelkar But why to add new argument when we can do work without it? Also, no matter from where you call close_server, we have to do remove_ptr_from_array in it so I will keep current code as it.

Reply:
In general it is not good to alter the array which is referenced a while loop from a function which is called inside the same loop i.e. we are basically altering(removing elements from the array) the array within the same loop. Although here it does not harm because we are closing for all the servers. Secondly we can avoid calling count_array() in loop. 

This is just another way of doing it. Having said that I am fine if you want to leave the code as is.

Reply:
Thanks @suresh-thelkar for information. I will keep code as it.

Reply:
Since we know close_server() will decrement the size of servers, why don't we set count_array() into a variable and decrement it.  The way you currently have it, count_array() is an O(N^2) loop.  It only needs to be O(N)

Reply:
Even better I can do like:
```c
while (servers && servers[0])
   close_server(severs[0]);
```
As while closing last server, close_server will set servers = NULL. I will update code.